WoodStar/WoodStar/TdsChannel.cs

```diff
@@ -0,0 +1,156 @@
+            preloginStream.WriteToBuffer(preloginBuffer);
+
+            await _stream.WriteAsync(preloginBuffer, 0, header.Length);
+            await _stream.FlushAsync();
+
+            ArrayPool<byte>.Shared.Return(preloginBuffer);
+
+            var preloginResponse = await ReadNextPacketAsync(ResponseType.Prelogin);
+            if (preloginResponse is PreloginPacket preloginPacket)
+            {
+                // Assign version etc info to connection
+            }
+        }
+
+        public async Task Login(string username, string password, string database)
+        {
+            var login7Stream = new Login7Stream
+            {
+                UserName = username,
+                Password = password,
+                Database = database,
+            };
+            var header = new TdsHeader(PacketType.Login, PacketStatus.EOM, login7Stream.Length + TdsHeader.HeaderSize, 0, 1);
+            var loginBuffer = ArrayPool<byte>.Shared.Rent(header.Length);
+            header.WriteToBuffer(loginBuffer);
+            login7Stream.WriteToBuffer(loginBuffer);
+
+            await _sslStream.WriteAsync(loginBuffer, 0, header.Length);
+            await _sslStream.FlushAsync();
+
+            ArrayPool<byte>.Shared.Return(loginBuffer);
+
+            await ReadNextPacketAsync(ResponseType.Login);
+        }
+
+        public async Task ExecuteSqlAsync(string sql)
+        {
+            var sqlBatchStream = new SqlBatchStream(
+                new StreamHeader
+                {
+                    TransactionDescriptor = new TransactionDescriptorHeader(0, 1)
+                },
+                sql);
+
+            var header = new TdsHeader(PacketType.SQLBatch, PacketStatus.EOM, sqlBatchStream.Length + TdsHeader.HeaderSize, 0, 1);
+            var sqlBatchBuffer = ArrayPool<byte>.Shared.Rent(header.Length);
+            header.WriteToBuffer(sqlBatchBuffer);
+            sqlBatchStream.WriteToBuffer(sqlBatchBuffer);
+            var b = string.Join(",", sqlBatchBuffer);
+            var c = string.Join(" ", sqlBatchBuffer.Select(e => e.ToString("X2")));
+
+            await _sslStream.WriteAsync(sqlBatchBuffer, 0, header.Length);
+            await _sslStream.FlushAsync();
+
+            ArrayPool<byte>.Shared.Return(sqlBatchBuffer);
+
+            await ReadNextPacketAsync(ResponseType.SqlBatchResult);
+        }
+
+        private async ValueTask<TdsPacket?> ReadNextPacketAsync(ResponseType expectedResponse)
+        {
+            while (true)
+            {
+                var reader = expectedResponse == ResponseType.Prelogin || expectedResponse == ResponseType.Login
+                    ? _pipeReader
+                    : _sslPipeReader;
+                var result = await reader.ReadAtLeastAsync(8);
+                var buffer = result.Buffer;
+
+                if (buffer.Length >= TdsHeader.HeaderSize)
+                {
+                    var header = TdsHeader.Parse(buffer.Slice(0, TdsHeader.HeaderSize));
```

Comment:
What if we set the header into a variable, so we parse it only once per `ReadNextPacketAsync`?